### PR TITLE
dbeaver/pro#1902 Disable data export and copying data

### PR DESCRIPTION
**Pull request by serjiokov**
Provided policy data export-copy provider for restriction operation.

### CODE REVIEW

**Comment by serjiokov** [comment]
In order to restrict **export** data operation pls add in dbeaver.ini:
  -Dpolicy.data.export.disabled=true 
or in windows registry:
**HKEY_CURRENT_USER\Software\DBeaver Corp\DBeaver\policy**
 policy.data.export.disabled=true
 
In order to restrict **copy** data operation pls add in dbeaver.ini:
  -Dpolicy.data.copy.disabled=true 
or in windows registry:
**HKEY_CURRENT_USER\Software\DBeaver Corp\DBeaver\policy**
 policy.data.copy.disabled=true

**Comment by serjiokov** [comment]
Screenshot 2023-10-10 160604

**Comment by serjiokov** [comment]
Screenshot 2023-10-10 160514

**Review by E1izabeth — APPROVED** [review]

**Review by E1izabeth — APPROVED** [review]

File: plugins/org.jkiss.dbeaver.data.transfer.ui/src/org/jkiss/dbeaver/tools/transfer/ui/handlers/DataExportHandler.java [hunk 1]
```diff
@@ -34,4 +39,17 @@ protected IDataTransferNode<?> adaptTransferNode(Object object)
...
     }
 
+    @Override
+    public Object execute(ExecutionEvent event) throws ExecutionException {
+        BasePolicyDataProvider policyProvider = new BasePolicyDataProvider();
+        if (policyProvider.isExportDataDisabled()) {
+            MessageDialog.open(MessageDialog.WARNING,
+                HandlerUtil.getActiveShell(event),
+                DTUIMessages.dialog_policy_data_export_title,
+                DTUIMessages.dialog_policy_data_export_msg, 0);
```
Comment (serge-rider) [inline plugins/org.jkiss.dbeaver.data.transfer.ui/src/org/jkiss/dbeaver/tools/transfer/ui/handlers/DataExportHandler.java]: Use SWT. constants, never use magic numbers.

File: plugins/org.jkiss.dbeaver.ui/src/org/jkiss/dbeaver/runtime/policy/AbstractPolicyDataProvider.java [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.jkiss.dbeaver.runtime.policy;
+
+import org.eclipse.core.runtime.Platform;
+
+import com.sun.jna.platform.win32.Advapi32Util;
```
Comment (serge-rider) [inline plugins/org.jkiss.dbeaver.ui/src/org/jkiss/dbeaver/runtime/policy/AbstractPolicyDataProvider.java:19]: Wrong imports orders

File: plugins/org.jkiss.dbeaver.ui/src/org/jkiss/dbeaver/runtime/policy/AbstractPolicyDataProvider.java [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+            }
+            isPolicyEnabled = getBooleanFromWinRegitryNode(WinReg.HKEY_LOCAL_MACHINE, property);
+            if (isPolicyEnabled) {
+                return isPolicyEnabled;
+            }
+        }
+        return false;
+    }
+
+    private boolean getBooleanFromWinRegitryNode(WinReg.HKEY root, String property) {
```
Comment (serge-rider) [inline plugins/org.jkiss.dbeaver.ui/src/org/jkiss/dbeaver/runtime/policy/AbstractPolicyDataProvider.java]: Typo (regitry)

File: plugins/org.jkiss.dbeaver.ui/src/org/jkiss/dbeaver/runtime/policy/PolicyDataProvider.java [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+ */
+public interface PolicyDataProvider {
+
+    /**
+     * Retrieve policy value from system environment 
+     *
+     * @param propertyName - data policy
+     * @return - boolean policy value
+     */
+    boolean getDataPolicyFromSystem(String propertyName);
```
Comment (serge-rider) [inline plugins/org.jkiss.dbeaver.ui/src/org/jkiss/dbeaver/runtime/policy/PolicyDataProvider.java]: We don't need these function in interface. Leave just isDataPolicyEnabled.

File: plugins/org.jkiss.dbeaver.ui/src/org/jkiss/dbeaver/runtime/policy/AbstractPolicyDataProvider.java [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+        String isDisabledPropValue = System.getProperty(property);
+        if (isDisabledPropValue != null && !isDisabledPropValue.isEmpty()) {
+            return Boolean.valueOf(isDisabledPropValue);
+        }
+        return false;
+    }
+
+    @Override
+    public boolean getDataPolicyFromRegistry(String property) {
+        if (Platform.OS_WIN32.equals(Platform.getOS())) {
```
Comment (serge-rider) [inline plugins/org.jkiss.dbeaver.ui/src/org/jkiss/dbeaver/runtime/policy/AbstractPolicyDataProvider.java]: Use RuntimeUtils.isWindows

**Review by serge-rider — CHANGES_REQUESTED** [review]

File: plugins/org.jkiss.dbeaver.ui/src/org/jkiss/dbeaver/runtime/policy/AbstractPolicyDataProvider.java (hunk 5 above)
Comment (serjiokov) [reply plugins/org.jkiss.dbeaver.ui/src/org/jkiss/dbeaver/runtime/policy/AbstractPolicyDataProvider.java]: Done

File: plugins/org.jkiss.dbeaver.ui/src/org/jkiss/dbeaver/runtime/policy/PolicyDataProvider.java (hunk 4 above)
Comment (serjiokov) [reply plugins/org.jkiss.dbeaver.ui/src/org/jkiss/dbeaver/runtime/policy/PolicyDataProvider.java]: Done

File: plugins/org.jkiss.dbeaver.ui/src/org/jkiss/dbeaver/runtime/policy/AbstractPolicyDataProvider.java (hunk 3 above)
Comment (serjiokov) [reply plugins/org.jkiss.dbeaver.ui/src/org/jkiss/dbeaver/runtime/policy/AbstractPolicyDataProvider.java]: Done

File: plugins/org.jkiss.dbeaver.data.transfer.ui/src/org/jkiss/dbeaver/tools/transfer/ui/handlers/DataExportHandler.java (hunk 1 above)
Comment (serjiokov) [reply plugins/org.jkiss.dbeaver.data.transfer.ui/src/org/jkiss/dbeaver/tools/transfer/ui/handlers/DataExportHandler.java]: Done

File: plugins/org.jkiss.dbeaver.ui.editors.data/src/org/jkiss/dbeaver/ui/controls/resultset/handler/ResultSetHandlerMain.java [hunk 6]
```diff
@@ -196,6 +197,7 @@ public Object execute(ExecutionEvent event) throws ExecutionException {
         String actionId = event.getCommand().getId();
         IResultSetPresentation presentation = rsv.getActivePresentation();
         DBPDataSource dataSource = rsv.getDataSource();
+        BasePolicyDataProvider policyProvider = new BasePolicyDataProvider();
```
Comment (serge-rider) [inline plugins/org.jkiss.dbeaver.ui.editors.data/src/org/jkiss/dbeaver/ui/controls/resultset/handler/ResultSetHandlerMain.java]: Ne need to create new instance here. This handler processes dozens of different commands.

File: plugins/org.jkiss.dbeaver.ui.editors.data/src/org/jkiss/dbeaver/ui/controls/resultset/handler/ResultSetHandlerMain.java [hunk 7]
```diff
@@ -411,9 +413,16 @@ public Object execute(ExecutionEvent event) throws ExecutionException {
                 break;
             }
             case IWorkbenchCommandConstants.EDIT_COPY:
-                ResultSetUtils.copyToClipboard(
-                    presentation.copySelection(
-                        new ResultSetCopySettings(false, false, false, true, false, null, null, null, DBDDisplayFormat.EDIT)));
+                if (policyProvider.isCopyDataDisabled()) {
```
Comment (serge-rider) [inline plugins/org.jkiss.dbeaver.ui.editors.data/src/org/jkiss/dbeaver/ui/controls/resultset/handler/ResultSetHandlerMain.java]: I don't think that we need to disable copy-paste completely.
Probably only when user copies multiple rows.

**Review by serge-rider — CHANGES_REQUESTED** [review]

**Comment by serge-rider** [comment]
Also I think we must hide export menu and toolbar items.

File: plugins/org.jkiss.dbeaver.ui.editors.data/src/org/jkiss/dbeaver/ui/controls/resultset/handler/ResultSetHandlerMain.java (hunk 6 above)
Comment (serjiokov) [reply plugins/org.jkiss.dbeaver.ui.editors.data/src/org/jkiss/dbeaver/ui/controls/resultset/handler/ResultSetHandlerMain.java]: Done. The handler instance moved to class field.

**Comment by Matvey16** [comment]
Verified

**Review by serge-rider — APPROVED** [review]

**Comment by serjiokov** [comment]
Reopened PR:  https://github.com/dbeaver/dbeaver/pull/21555

**Pull request merged**